### ISSUE TITLE
PHP 8 plugin build issue

### ISSUE THREAD

**Issue by jSML4ThWwBID69YC** [unbit/uwsgi#2283]
Hello, 

I'm trying to build the uWSGi plugin for PHP 8. I'm using the same steps I do for PHP 7, but with no success. 

OS: FreeBSD 12.2
PHP: PHP 8.0.1 from the ports tree with the following basic options set. 

-  ZTS: Disabled 
-  Embed: Enabled
-  CLI: Enabled

uWSGI version: 2.0.19.1

Build commands:
Extract uWSGI 2.0.19.1.tar.gz to the /root/scripts/uwsgi folder. Then run:

```
cd /root/scripts/uwsgi
make PROFILE=nolang
./uwsgi --build-plugin "plugins/php php"

```

Error:
```
*** uWSGI building and linking plugin from plugins/php ***
[cc -pthread] php_plugin.so
plugins/php/php_plugin.c:77:70: error: expected ')'
static size_t sapi_uwsgi_ub_write(const char *str, size_t str_length TSRMLS_DC)
                                                                     ^
plugins/php/php_plugin.c:77:34: note: to match this '('
static size_t sapi_uwsgi_ub_write(const char *str, size_t str_length TSRMLS_DC)
                                 ^
plugins/php/php_plugin.c:92:70: error: expected ')'
static int sapi_uwsgi_send_headers(sapi_headers_struct *sapi_headers TSRMLS_DC)
                                                                     ^
plugins/php/php_plugin.c:92:35: note: to match this '('
static int sapi_uwsgi_send_headers(sapi_headers_struct *sapi_headers TSRMLS_DC)
                                  ^
plugins/php/php_plugin.c:127:69: error: expected ')'
static size_t sapi_uwsgi_read_post(char *buffer, size_t count_bytes TSRMLS_DC)
                                                                    ^
plugins/php/php_plugin.c:127:35: note: to match this '('
static size_t sapi_uwsgi_read_post(char *buffer, size_t count_bytes TSRMLS_DC)
                                  ^
plugins/php/php_plugin.c:167:66: error: expected ')'
static void sapi_uwsgi_register_variables(zval *track_vars_array TSRMLS_DC)
                                                                 ^
plugins/php/php_plugin.c:167:42: note: to match this '('
static void sapi_uwsgi_register_variables(zval *track_vars_array TSRMLS_DC)
                                         ^
plugins/php/php_plugin.c:171:52: error: expected ')'
        php_import_environment_variables(track_vars_array TSRMLS_CC);
                                                          ^
plugins/php/php_plugin.c:171:34: note: to match this '('
        php_import_environment_variables(track_vars_array TSRMLS_CC);
                                        ^
plugins/php/php_plugin.c:175:114: error: expected ')'
                php_register_variable_safe("SERVER_SOFTWARE", uphp.server_software, uphp.server_software_len, track_vars_array TSRMLS_CC);
                                                                                                                               ^
plugins/php/php_plugin.c:175:29: note: to match this '('
                php_register_variable_safe("SERVER_SOFTWARE", uphp.server_software, uphp.server_software_len, track_vars_array TSRMLS_CC);
                                          ^
plugins/php/php_plugin.c:178:78: error: expected ')'
                php_register_variable_safe("SERVER_SOFTWARE", "uWSGI", 5, track_vars_array TSRMLS_CC);
                                                                                           ^
plugins/php/php_plugin.c:178:29: note: to match this '('
                php_register_variable_safe("SERVER_SOFTWARE", "uWSGI", 5, track_vars_array TSRMLS_CC);
                                          ^
plugins/php/php_plugin.c:184:21: error: expected ')'
                        track_vars_array TSRMLS_CC);
                                         ^
plugins/php/php_plugin.c:182:29: note: to match this '('
                php_register_variable_safe( estrndup(wsgi_req->hvec[i].iov_base, wsgi_req->hvec[i].iov_len),
                                          ^
plugins/php/php_plugin.c:187:105: error: expected ')'
        php_register_variable_safe("PATH_INFO", wsgi_req->path_info, wsgi_req->path_info_len, track_vars_array TSRMLS_CC);
                                                                                                               ^
plugins/php/php_plugin.c:187:28: note: to match this '('
        php_register_variable_safe("PATH_INFO", wsgi_req->path_info, wsgi_req->path_info_len, track_vars_array TSRMLS_CC);
                                  ^
plugins/php/php_plugin.c:189:115: error: expected ')'
                php_register_variable_safe("QUERY_STRING", wsgi_req->query_string, wsgi_req->query_string_len, track_vars_array TSRMLS_CC);
                                                                                                                                ^
plugins/php/php_plugin.c:189:29: note: to match this '('
                php_register_variable_safe("QUERY_STRING", wsgi_req->query_string, wsgi_req->query_string_len, track_vars_array TSRMLS_CC);
                                          ^
plugins/php/php_plugin.c:192:111: error: expected ')'
        php_register_variable_safe("SCRIPT_NAME", wsgi_req->script_name, wsgi_req->script_name_len, track_vars_array TSRMLS_CC);
                                                                                                                     ^
plugins/php/php_plugin.c:192:28: note: to match this '('
        php_register_variable_safe("SCRIPT_NAME", wsgi_req->script_name, wsgi_req->script_name_len, track_vars_array TSRMLS_CC);
                                  ^
plugins/php/php_plugin.c:193:101: error: expected ')'
        php_register_variable_safe("SCRIPT_FILENAME", wsgi_req->file, wsgi_req->file_len, track_vars_array TSRMLS_CC);
                                                                                                           ^
plugins/php/php_plugin.c:193:28: note: to match this '('
        php_register_variable_safe("SCRIPT_FILENAME", wsgi_req->file, wsgi_req->file_len, track_vars_array TSRMLS_CC);
                                  ^
plugins/php/php_plugin.c:195:117: error: expected ')'
        php_register_variable_safe("DOCUMENT_ROOT", wsgi_req->document_root, wsgi_req->document_root_len, track_vars_array TSRMLS_CC);
                                                                                                                           ^
plugins/php/php_plugin.c:195:28: note: to match this '('
        php_register_variable_safe("DOCUMENT_ROOT", wsgi_req->document_root, wsgi_req->document_root_len, track_vars_array TSRMLS_CC);
                                  ^
plugins/php/php_plugin.c:202:130: error: expected ')'
                php_register_variable_safe("PATH_TRANSLATED", path_translated, wsgi_req->file_len + wsgi_req->path_info_len , track_vars_array TSRMLS_CC);
                                                                                                                                               ^
plugins/php/php_plugin.c:202:29: note: to match this '('
                php_register_variable_safe("PATH_TRANSLATED", path_translated, wsgi_req->file_len + wsgi_req->path_info_len , track_vars_array TSRMLS_CC);
                                          ^
plugins/php/php_plugin.c:205:73: error: expected ')'
                php_register_variable_safe("PATH_TRANSLATED", "", 0, track_vars_array TSRMLS_CC);
                                                                                      ^
plugins/php/php_plugin.c:205:29: note: to match this '('
                php_register_variable_safe("PATH_TRANSLATED", "", 0, track_vars_array TSRMLS_CC);
                                          ^
plugins/php/php_plugin.c:208:108: error: expected ')'
        php_register_variable_safe("PHP_SELF", wsgi_req->script_name, wsgi_req->script_name_len, track_vars_array TSRMLS_CC);
                                                                                                                  ^
plugins/php/php_plugin.c:208:28: note: to match this '('
        php_register_variable_safe("PHP_SELF", wsgi_req->script_name, wsgi_req->script_name_len, track_vars_array TSRMLS_CC);
                                  ^
plugins/php/php_plugin.c:215:48: error: expected ')'
                                equal+1, strlen(equal+1), track_vars_array TSRMLS_CC);
                                                                           ^
plugins/php/php_plugin.c:214:30: note: to match this '('
                        php_register_variable_safe( estrndup(usl->value, equal-usl->value),
                                                  ^
plugins/php/php_plugin.c:259:101: error: expected ')'
                        zend_register_string_constant(name, name_len, strval, CONST_CS | CONST_PERSISTENT, module_number TSRMLS_CC);
                                                                                                                         ^
plugins/php/php_plugin.c:259:33: note: to match this '('
                        zend_register_string_constant(name, name_len, strval, CONST_CS | CONST_PERSISTENT, module_number TSRMLS_CC);
                                                     ^
plugins/php/php_plugin.c:292:51: error: expected ')'
        if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "s|s", &key, &keylen, &cache, &cachelen) == FAILURE) {
                                                  ^
plugins/php/php_plugin.c:292:34: note: to match this '('
        if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "s|s", &key, &keylen, &cache, &cachelen) == FAILURE) {
                                 ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
*** unable to build php plugin ***
```

**Comment by xrmx** [comment]
Could you please check that master is working fine?

**Comment by dvzrv** [comment]
@xrmx I have tested this with php 8.0.2 now (using nextcloud as my guinea pig) and it works.

The only thing I am currently still trying to figure out is, how running a plugin for php7 and php8 side by side will work. Do you have any recommendations?

When using nginx the usual way of addressing the specific uwsgi plugin seems to be via `uwsgi_modifier1 14;`. How does this behave if I have a php7 and a php (8) plugin side-by-side? Does the first one found get loaded by uwsgi? Currently they both have the same modifier as [I only applied the minimum amount of modifications](https://github.com/archlinux/svntogit-community/blob/00bc8ab4321286cfcaad41849996fbfd66a08f71/trunk/PKGBUILD#L112-L122) I could think of.

However, I guess [I have to patch this line](https://github.com/unbit/uwsgi/blob/793bf9392122512855146b61539f5cc8a7572376/plugins/php/php_plugin.c#L1131) to assign a different number for e.g. the php7 plugin (as long as we need it). Sorry for my ignorance of any standards there... what would be a good number? Are there standards to adhere to?

**Comment by xrmx** [comment]
You should be able to override the modifier in configuration, e.g. `plugins = http,0:php` and you can build the same plugin with different names, see `.github/workflows/test.yml`.

**Comment by jSML4ThWwBID69YC** [comment]
> Could you please check that master is working fine?

@xrmx 
The Master branch can't find -lphp8.

```
*** uWSGI building and linking plugin from plugins/php ***
[cc -pthread] php_plugin.so
ld: error: unable to find library -lphp8
cc: error: linker command failed with exit code 1 (use -v to see invocation)
*** unable to build php plugin ***
```

It appears on FreeBSD, the library is no longer versioned. It's now ``` /usr/local/lib/libphp.so ``` instead of ```/usr/local/lib/libphp8.so```. Symlinking libphp.so to libphp8.so works around it. I'll give it some serious testing this weekend, but so far it all looks workable.

**Comment by dvzrv** [comment]
> You should be able to override the modifier in configuration, e.g. plugins = http,0:php and you can build the same plugin with different names, see .github/workflows/test.yml.

@xrmx Could you elaborate on your suggestion a little bit? From the custom build system it does not become apparent how that would even be able to work, as [build_plugin()](https://github.com/unbit/uwsgi/blob/793bf9392122512855146b61539f5cc8a7572376/uwsgiconfig.py#L1434) does not separate the provided plugin string and [neither does the block calling it](https://github.com/unbit/uwsgi/blob/793bf9392122512855146b61539f5cc8a7572376/uwsgiconfig.py#L547-L556).
Using your above example would lead to:

```
*** building plugin: 0:php7 ***
Error: unable to find directory 'plugins/0:php7'
```

The github ci setup [has examples for using different python interpreters](https://github.com/unbit/uwsgi/blob/793bf9392122512855146b61539f5cc8a7572376/.github/workflows/test.yml#L31-L50), but no test setup for different php interpreters/versions.

Additionally, this does not answer the question in regards to the assigned modifier1 number. Is e.g. 14 for the php plugin chosen by the uwsgi developers, or does it follow a specific schema, that one needs to be aware of?

> It appears on FreeBSD, the library is no longer versioned. It's now /usr/local/lib/libphp.so instead of /usr/local/lib/libphp8.so. Symlinking libphp.so to libphp8.so works around it. I'll give it some serious testing this weekend, but so far it all looks workable.

@jSML4ThWwBID69YC It is similar on Arch Linux and I removed the hardcoded concatenation of the version suffix before building the plugin (see the link in [my previous comment](https://github.com/unbit/uwsgi/issues/2283#issuecomment-786814869)).

**Comment by xrmx** [comment]
@dzrv I think you conflated the two things: 

```
$ python3 uwsgiconfig.py --plugin plugins/php base php8
using profile: buildconf/base.ini
detected include path: ['/usr/lib/gcc/x86_64-linux-gnu/10/include', '/usr/local/include', '/usr/include/x86_64-linux-gnu', '/usr/include']
*** uWSGI building and linking plugin plugins/php ***
[x86_64-linux-gnu-gcc -pthread] ./php8_plugin.so
build time: 0 seconds
*** php8 plugin built and available in ./php8_plugin.so ***
```

And I guess you can tune the different php version with environment variables. Then you can remap the modifier when *loading* the plugin in uwsgi configuration.

**Comment by dvzrv** [comment]
> And I guess you can tune the different php version with environment variables

That is only possible to a certain degree. The plugin configuration uses `php --version` to [derive PHP's major version](https://github.com/unbit/uwsgi/blob/793bf9392122512855146b61539f5cc8a7572376/plugins/php/uwsgiplugin.py#L15) and later [appends](https://github.com/unbit/uwsgi/blob/793bf9392122512855146b61539f5cc8a7572376/plugins/php/uwsgiplugin.py#L24) the suffix during linking. On Arch Linux the current php is 8.0.2, but the exposed library (via the [php-embed](https://archlinux.org/packages/extra/x86_64/php-embed/) package) is `/usr/lib/libphp.so` not `/usr/lib/libphp8.so`. I can not build the php plugin (using the current php) without modifying the plugin configuration.
However, your suggestion of using `uwsgiconfig.py --plugin <plugin_dir> <base> <name>` at least saves some patchwork on the php7 version! Thanks :)

> Then you can remap the modifier when loading the plugin in uwsgi configuration.

Could you be a bit more specific on what you mean by that? How/ in what context would one remap the modifier1? By [patching php_plugin.c](https://github.com/unbit/uwsgi/blob/793bf9392122512855146b61539f5cc8a7572376/plugins/php/php_plugin.c#L1131) or by some other means? If this can be done via patching: What modifier1 uint8 is a good one (e.g. could I just assign the php7 plugin the number 13 instead and refer to that when using it in nginx)?
If it can be done by other means, please explain! :)

**Comment by dvzrv** [comment]
> > Then you can remap the modifier when loading the plugin in uwsgi configuration.
> 
> Could you be a bit more specific on what you mean by that? How/ in what context would one remap the modifier1? By [patching php_plugin.c](https://github.com/unbit/uwsgi/blob/793bf9392122512855146b61539f5cc8a7572376/plugins/php/php_plugin.c#L1131) or by some other means? If this can be done via patching: What modifier1 uint8 is a good one (e.g. could I just assign the php7 plugin the number 13 instead and refer to that when using it in nginx)?
> If it can be done by other means, please explain! :)

To answer my own question, it seems that reusing the `modifier1 = 14` is no problem (for php7 and php(8) side-by-side).
One only has to specify e.g. `plugins = php7` in the uwsgi configuration file (or commandline argument) of a software that still relies on php7 (and I guess this was what @xrmx was referring to).

**Comment by xrmx** [comment]
I thought you wanted to load both PHP plugins at the same time, and so
suggested you can give to one of them a different modifier which should be
unique. As you found out one of a time there's no special configuration
needed.

Il sab 27 feb 2021, 13:26 David Runge <notifications@github.com> ha scritto:

**Comment by dvzrv** [comment]
> I thought you wanted to load both PHP plugins at the same time,

I do.

> you can give to one of them a different modifier which should be
unique. As you found out one of a time there's no special configuration
needed.

I have just tested the following: One php7 application (using the php7 plugin) and one php(8) application (using the php(8) plugin), side-by-side. Both plugins rely on the modifier1 being set to `14`, the uwsgi configuration for the php7 application specifies `plugins = php7`, whereas the uwsgi configuration for the php(8) application specifies `plugins = php`. Both work simultaneously without issues (AFAICS).

Are you sure the `modifier1` needs to be adjusted by patching the plugin source code for e.g. the php7 plugin? It seems to work by just specifying the correct plugin in the uwsgi configuration of the respective application.

Thanks for the feedback! :)

**Comment by xrmx** [comment]
@dvzrv With "at the same time" i mean at runtime as in both plugins loaded in the same uwsgi instance.

**Comment by dvzrv** [comment]
> @dvzrv With "at the same time" i mean at runtime as in both plugins loaded in the same uwsgi instance.

Ahh, okay. Thanks for clearing that up!

I believe for the packages in our repositories, that still rely on php7 the preferred way of starting them is with one uwsgi configuration each (as they are ideally run as their own user each, for privilege separation).
I think I will therefore not introduce a separate `modifier1` for the php7 plugin.

Out of curiosity though and in case it is needed in the future: Is there a specific schema to adhere to, when assigning a new `modifier1` for a plugin (apart from it being unique of course), or is it basically free-form?

**Comment by xrmx** [comment]
Yes, I'd say don't poke with modifier as a distro package level. Only users that have very special needs should confront the complication.

**Comment by morsik** [comment]
The same problem happens on CentOS 7 with PHP 8.0 from Remi Repository :(

**Comment by xrmx** [comment]
Build isses should be fixed in 2.0.21.

**Comment by remicollet** [comment]
I don't understand why this one is closed

Please see pr #2581
